lib/golem/doc/system-design.md

```diff
@@ -0,0 +1,80 @@
+# Arm control system design
+
+We want to achieve the following with our arm control system:
+
+- Track the state (position, velocity, torque) of the arm
+- Move the arm to desired positions while satisfying constraints:
+    * Avoid obstacles
+    * Limit maximum velocity and torque on the joints
+- Detect when the arm is blocked
+
+For that, we'll get inspiration from [Mikrob technology's control system design in aversive](https://wiki.droids-corp.org/articles/a/v/e/Aversive/Asservissement_Microb_2008.html).
+It provides us with a modular clean architecture that separates concerns properly.
+The separate components can be composed to implement complex behavior, and some components are optional.
+Here follows a diagram of the proposed design.
+
+![Diagram of the proposed arm control system design](figures/system.pdf)
+
+Each of the main blocks (motors, control loop, trajectory, and blocking detection) run in their own thread.
+In the following, their respective role is described in more detail.
+
+## Motors
+
+The motors are interfaced over UAVCAN.
+The current abstraction goes through `motor_driver` and `motor_manager` in `master-firmware`.
+This allows us to:
+
+- Gather feedback on the motors: we'll use position, velocity, torque
+- Send setpoints to the motors: we'll use torque
+
+This part already exists.
+
+## Control loop
+
+This block is broken down into three major components:
+
+- State estimator
+- Consign filter
+- Control law
+
+### State estimator
+
+The state estimator is responsible for gathering sensor measurements (i.e. joint positions, velocities, and torques) and estimate the state of the arm.
+The simplest possible state estimator only knows about the arm's kinematics and dynamics, such that it can convert from joint space to cartesian space, and viceversa.
+A more elaborate state estimator can be a Kalman filter which attempts to model the presence of noise in the input joint measurements and filter it out.
+
+### Consign filter
+
+The consign filter is an optional block.
+It takes a consign (desired target arm state) and smoothes it, usually in order to make it feasible.
+It can be a limited acceleration filter (ramp) or a limited velocity filter (quadramp).
```

Comment:
If I am not mistaken, at least in Aversive the ramp is limited velocity, while quadramp limits accel + speed.

Reply:
Also, how do you handle joint space limits such as torque limits ?

Reply:
I'd say that's for the controller to manage (e.g. MPC). The consign filter may be discarded in some cases.